test/Calc.mo

```diff
@@ -0,0 +1,103 @@
+import A "mo:adapton/adapton";
+import D "mo:adapton/draw";
+import E "mo:adapton/evalType";
+import H "mo:base/hash";
+import L "mo:base/list";
+import R "mo:base/result";
+
+module {
+
+public type Name = Text;
+
+public type Val = Int;
+
+public type Error = {
+  #divByZero;
+  #unimplemented;
+  #putError : Name;
+};
+
+public type Exp = {
+  #num: Int;
+  #add: (Exp, Exp);
+  #sub: (Exp, Exp);
+  #mul: (Exp, Exp);
+  #div: (Exp, Exp);
+  #named: (Name, Exp); // record a cached result at Name
+};
+
+// simple integer-based calculator, with incremental caching
+class Calc() {
+
+  /* -- cache implementation, via adapton package -- */
+  var namedCache : A.Engine<Name, Val, Error, Exp> = {
+    let _errorEq = errorEq;
+    A.Engine<Name, Val, Error, Exp>({
+      nameEq=func (x:Text, y:Text) : Bool { x == y };
+      valEq=func (x:Int, y:Int) : Bool { x == y };
+      errorEq=_errorEq;
```

Comment:
@crusso @rossberg No matter where I put this "functor evaluation", I get a definedness error about the operations that it is passing to the "`Engine` functor" that it applies.

Reply:
It should be legal to do this, right?

I am referencing (not using) `eval` and the other operations here.

Reply:
The problem is that `eval` uses `namedCache` while the RHS of `namedCache` also passes `eval` to another function. The compiler cannot prove that this function does not eagerly call `eval`, at which point `namedCache` would not yet be initialised. So it has to reject the recursion.

It should work if you move `A.Engine` out of the RHS of `namedCache` and after the definitions of the other functions.

Reply:
> the compiler cannot prove that this function does not eagerly call eval, at which point namedCache would not yet be initialised. So it has to reject the recursion.

Yes, makes perfect sense.  It all became clear just after I left my desk yesterday and thought about it later that evening.  Thanks for confirming.

I am waiting for some lazy programmer to come out of the woodwork and say something here.  Perhaps they are still waiting for the right time ("laziness").

Thanks again.

Reply:
Fixed this by separating the initialization of the Engine into two parts:

One happens before the evaluation function is defined.  One happens later, which "back-patches" this definition into the engine so that the cycle is complete.